<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkGuidedProjectsIBMGPXX0CRSEN2517-2023-01-01">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo"  />
    </a>
</p>


# **Winning Techniques for Your Next Data Science Contest**


Estimated time needed: **60** minutes


The field of data science is rapidly growing, and the need for individuals in the traditional industries, such as Agriculture, Transportation, Construction, Retail, Hospitality and Tourism, to acquire these skills is also increasing. If you are an expert in these industries and would like to have a competitive edge in your career and the ability to drive innovation in your industry, then you found the right place!


This guided project provides a unique opportunity for you to learn about data science and machine learning in a practical way. We will guide you through the full machine learning project cycle, which typically includes the following steps:

- **Problem formulation**: defining the problem and understanding the business context.

- **Data collection**: gathering and acquiring data from various sources.

- **Data preprocessing**: cleaning, transforming, and preparing data for analysis.

- **Exploratory data analysis**: understanding the structure, contents, and patterns in the data through visualizations and summary statistics.

- **Feature engineering**: creating new features from the existing data.

- **Model selection**: choosing an appropriate model or algorithm for the problem.

- **Model training**: fitting the model to the data and tuning its hyperparameters.

- **Model evaluation**: assessing the performance of the model using appropriate metrics and techniques.


This comprehensive guide is hard to find on the internet and will equip you with a solid foundation in the data science process.

After completing the project, you will be able to use real-world data to make predictions and apply machine learning algorithms to analyze and draw insights from data.



<center><img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-GPXX0CRSEN/images/Screen%20Shot%202023-04-03%20at%2010.51.22%20PM.png" width="60%"></center>

<p style="color:gray; text-align:center;">Credit: THINKSTOCK</p>


## __Table of Contents__

<ol>
    <li><a href="#Objectives">Objectives</a></li>
    <li>
        <a href="#Setup">Setup</a>
        <ol>
            <li><a href="#Installing-Required-Libraries">Installing Required Libraries</a></li>
            <li><a href="#Importing-Required-Libraries">Importing Required Libraries</a></li>
            <li><a href="#Defining-Helper-Functions">Defining Helper Functions</a></li>
        </ol>
    </li>
    <li><a href="#Background-(optional)">Background (optional)</a></li>
    <li><a href="#Loading the dataset">Loading the dataset</a></li>
    <li>
        <a href="#Preprocessing the dataset">Preprocessing the dataset</a>
        <ol>
            <li><a href="#Make Loc Groups">Make Loc Groups</a></li>
            <li><a href="#Fill Missing Values">Fill Missing Values</a></li>
        </ol>
    </li>
    <li>
        <a href="#Feature Selection and Lasso Regressions (to Obtain Residuals)">Feature Selection and Lasso Regressions (to Obtain Residuals)</a>
        <ol>
            <li><a href="#Train, Validation, Test Split">Train, Validation, Test Split</a></li>
            <li><a href="#Feature Selection For Linear Regression">Feature Selection For Linear Regression</a></li>
            <li><a href="#Selecting Alpha and Fitting Linear Regressions (to Obtain Residuals)">Selecting Alpha and Fitting Linear Regressions (to Obtain Residuals)</a></li>
        </ol>
    </li>
    <li>
        <a href="#Using Boosting Algorithms to Predict Residuals">Using Boosting Algorithms to Predict Residuals</a>
        <ol>
            <li><a href="#Feature Selection for Boosting">Feature Selection for Boosting</a></li>
            <li><a href="#Boosting Models for Predicting Fitted Residuals">Boosting Models for Predicting Fitted Residuals</a></li>
        </ol>
    </li>
</ol>    


## Objectives

After completing this lab you will be able to:

- Learn the full machine learning project cycle, from data exploration to model evaluation
- Gain hands-on experience with data cleaning and preparation techniques
- Understand the importance of feature engineering and how to apply it to real-world datasets
- Learn how to select and train machine learning models for different problems
- Learn how to make predictions and draw insights from data using machine learning algorithms
- Understand the working mechanisms of three famous boosting algorithms: CatBoost, LightGBM, and XGBoost.
- Discover how to use ensemble learning techniques to combine model predictions and achieve even better results.


----


## Setup


For this lab, we will be using the following libraries:

*   [`pandas`](https://pandas.pydata.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMML0187ENSkillsNetwork31430127-2021-01-01) for managing the data.
*   [`numpy`](https://numpy.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMML0187ENSkillsNetwork31430127-2021-01-01) for mathematical operations.
*   [`sklearn`](https://scikit-learn.org/stable/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMML0187ENSkillsNetwork31430127-2021-01-01) for machine learning and machine-learning-pipeline related functions.
*   [`seaborn`](https://seaborn.pydata.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMML0187ENSkillsNetwork31430127-2021-01-01) for visualizing the data.
*   [`matplotlib`](https://matplotlib.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMML0187ENSkillsNetwork31430127-2021-01-01) for additional plotting tools.


### Installing Required Libraries

The following required libraries are pre-installed in the Skills Network Labs environment. However, if you run this notebook commands in a different Jupyter environment (e.g. Watson Studio or Ananconda), you will need to install these libraries by removing the `#` sign before `!mamba` in the code cell below.


In [ ]:
# All Libraries required for this lab are listed below. The libraries pre-installed on Skills Network Labs are commented.
# !mamba install -qy pandas==1.3.4 numpy==1.21.4 seaborn==0.9.0 matplotlib==3.5.0 scikit-learn==0.20.1
# Note: If your environment doesn't support "!mamba install", use "!pip install"

The following required libraries are __not__ pre-installed in the Skills Network Labs environment. __You will need to run the following cell__ to install them:


In [1]:
!mamba install -qy catboost
!mamba install -qy lightgbm
!mamba install -qy xgboost

Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... 

    Installed package of scikit-learn can be accelerated using scikit-learn-intelex.
    More details are available here: https://intel.github.io/scikit-learn-intelex

    For example:

        $ conda install scikit-learn-intelex
        $ python -m sklearnex my_application.py

    

done
Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done


### Importing Required Libraries

_We recommend you import all required libraries in one place (here):_


In [2]:
import warnings
warnings.simplefilter('ignore')

import pandas as pd
import numpy as np
import tqdm
from numpy import median
import random
import catboost as cb
import lightgbm as lgb
import xgboost as xgb
from sklearn.decomposition import PCA
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler
from sklearn.linear_model import Lasso, Ridge
from sklearn.ensemble import RandomForestClassifier
from statsmodels.stats.outliers_influence import variance_inflation_factor

import matplotlib.pyplot as plt
from sklearn.metrics import mean_squared_error
from sklearn.metrics import roc_auc_score
%matplotlib inline

### Defining Helper Functions


In [3]:
def rmse(actual, predicted):
    return mean_squared_error(actual, predicted, squared=False)

## Background


Participating in data science competitions is a great way for beginners to gain practical experience and develop valuable data science skills. Competitions provide real-world datasets and problems that allow you to apply your knowledge and practice the entire data science process.

The dataset we will use in this project is provided by Women in Data Science (WiDS Datathon) 2023 competition on Kaggle. This compeition focuses on forecasting weather in different regions within the US over a two-week period. Each row in the data corresponds to a single location and a single start date for the two-week period. **Our task is to predict the average temperature over the next 14 days, for each location and start date**.

The competition is now finished and it had over 700 participating teams around the world. We collected the best winning techniques of the top winners of the compeition and built this project. We offer you this great opportunity to learn from the best approaches and elevate your data science skills!


<center><img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-GPXX0CRSEN/images/bg.png" width="85%"></center>


In a nutshell, in this project we will first preprocess the dataset by filling the missing values and encode valuable features. Then we will use Lasso regression to predict the residuals of our original target, which allows us to capture any simple, linear relationships existing between the predictors and the target. After we obtain the fitted residuals, we will bring on the power of Boosting algorithms, which we will introduce later in this project, to predict the fitted residuals. Boosting algorithms will help us capture complex, non-linear pattern in our data, and we hope to further reduce the gap between our predictions and the real target using Boosting algorithms.

There are a lot of great technical details covered in this comprehensive project, so let's get started!


## Loading the dataset


We have a pre-prepared dataset consisting of features such as weather and climate information for a number of US locations, for a number of start dates for the two-week observation, as well as the forecasted temperature and precipitation from a number of weather forecast models. For more information regarding the data schema, please refer to the [competition](https://www.kaggle.com/competitions/widsdatathon2023/data?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkGuidedProjectsIBMGPXX0CRSEN2517-2023-01-01).

Let's import the dataset. It may take 2-3 minutes to import as the original dataset is quite large.


In [4]:
df = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-GPXX0CRSEN/LargeData/train_data.csv", parse_dates=['startdate'], index_col=0)
df.head()

lat       lon  startdate  contest-pevpr-sfc-gauss-14d__pevpr  \
index                                                                 
0      0.0  0.833333 2014-09-01                              237.00   
1      0.0  0.833333 2014-09-02                              228.90   
2      0.0  0.833333 2014-09-03                              220.69   
3      0.0  0.833333 2014-09-04                              225.28   
4      0.0  0.833333 2014-09-05                              237.24   

       nmme0-tmp2m-34w__cancm30  nmme0-tmp2m-34w__cancm40  \
index                                                       
0                         29.02                     31.64   
1                         29.02                     31.64   
2                         29.02                     31.64   
3                         29.02                     31.64   
4                         29.02                     31.64   

       nmme0-tmp2m-34w__ccsm30  nmme0-tmp2m-34w__ccsm40  \
index                                                     
0                        29.57                    30.73   
1                        29.57                    30.73   
2                        29.57                    30.73   
3                        29.57                    30.73   
4                        29.57                    30.73   

       nmme0-tmp2m-34w__cfsv20  nmme0-tmp2m-34w__gfdlflora0  ...  \
index                                                        ...   
0                        29.71                        31.52  ...   
1                        29.71                        31.52  ...   
2                        29.71                        31.52  ...   
3                        29.71                        31.52  ...   
4                        29.71                        31.52  ...   

       wind-vwnd-925-2010-11  wind-vwnd-925-2010-12  wind-vwnd-925-2010-13  \
index                                                                        
0                     -27.68                 -37.21                   8.32   
1                     -21.13                 -36.57                   8.77   
2                     -10.72                 -34.16                   6.99   
3                       0.33                 -31.04                   6.17   
4                       9.83                 -31.80                   7.47   

       wind-vwnd-925-2010-14  wind-vwnd-925-2010-15  wind-vwnd-925-2010-16  \
index                                                                        
0                       9.56                  -2.03                  48.13   
1                      21.17                   4.44                  48.60   
2                      32.16                   5.01                  48.53   
3                      39.66                  -1.41                  50.59   
4                      38.62                  -5.21                  54.73   

       wind-vwnd-925-2010-17  wind-vwnd-925-2010-18  wind-vwnd-925-2010-19  \
index                                                                        
0                      28.09                 -13.50                  11.90   
1                      27.41                 -23.77                  15.44   
2                      19.21                 -33.16                  15.11   
3                       8.29                 -37.22                  18.24   
4                      -2.58                 -42.30                  21.91   

       wind-vwnd-925-2010-20  
index                         
0                       4.58  
1                       3.42  
2                       4.82  
3                       9.74  
4                      10.95  

[5 rows x 245 columns]

In [5]:
df.shape

(375734, 245)

In [6]:
target = 'contest-tmp2m-14d__tmp2m'
df[target]

index
0         28.744480
1         28.370585
2         28.133059
3         28.256798
4         28.372353
            ...    
375729    17.150954
375730    16.962051
375731    16.915474
375732    16.536761
375733    15.910995
Name: contest-tmp2m-14d__tmp2m, Length: 375734, dtype: float64

As you can see, the original dataset contains 375,734 observations, 244 features and one target variable `contest-tmp2m-14d__tmp2m`. According to the competition, `contest-tmp2m-14d__tmp2m` is the **arithmetic mean of the max and min observed temperature over the next 14 days for each location and start date**.


## Preprocessing the dataset

This section explains how we preprocessed the original dataset and then split it into train, validation, and test datasets. Preprocessing is a critical step in building a successful Machine Learning project.


### Make Location Groups

First, the location data given are expressed in terms of longitude and latitude, we can turn this critical information into a more convenient format by creating location groups. A location group has its unique combination of longitude and latitude. However, to avoid having too many different groups, we choose to round the longitudes and latitudes down to contain 14 decimal places. 


In [7]:
def loc_group(df):
    
    scale = 14

    df['lat'] = np.round(df['lat'], scale)
    df['lon'] = np.round(df['lon'], scale)
    df['loc_group'] = df.groupby(['lon', 'lat']).ngroup()
     
    return df

In [8]:
df = loc_group(df)
df.loc_group.nunique()

514

We created in total 514 different location groups for which we will predict the average weather in the next 14 days.


In [9]:
n_startdates = []

for n in range(df.loc_group.nunique()):
    n_startdates.append(df[df.loc_group == n].startdate.nunique())

print(n_startdates)
print()
print(df.startdate.nunique())

[731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731, 731,

We can see that each `loc_group` has 731 unique start dates and we have in total 731 unique start dates in the entire dataset, meaning all location groups have data for the same 731 start dates. This is a good thing because we have a balanced dataset for training models to forecast weather in different locations.  


### Fill Missing Values


Let's examine columns with missing values in the original dataset.


In [ ]:
na_cols = []

for col in df.columns:
    if df[col].isnull().sum() >0:
        na_cols.append(col)
        print(col, df[col].isnull().sum())

According to the competition, **cancm30, cancm40, ccsm30, ccsm40, cfsv20, gfdlflora0, gfdlflorb0, gfdl0, nasa0, and nmme0mean** are most recent forecasts from weather models. 

- `nmme0-prate-34w`: weeks 3-4 weighted average of most recent monthly NMME model forecasts for precipitation.
- `nmme0-prate-56w`: weeks 5-6 weighted average of most recent monthly NMME model forecasts for precipitation.
- `nmme-prate-34w`: weeks 3-4 weighted average of monthly NMME model forecasts for precipitation.
- `nmme-prate-56w`: weeks 5-6 weighted average of monthly NMME model forecasts for precipitation.
- `nmme-tmp2m-34w`: weeks 3-4 weighted average of most recent monthly NMME model forecasts for target label, `contest-tmp2m-14d__tmp2m`.
- `nmme-tmp2m-56w`: weeks 5-6 weighted average of monthly NMME model forecasts for target label, `contest-tmp2m-14d__tmp2m`.



Now we know that we are missing forecasts from **ccsm30 and ccsm3** NMME (North American Multi-Model Ensemble) weather models for the above prefixes. 

We can make an educated guess that the missing forecasts can be derived from the remaining complete forecasts. Specifically, for column `nmme-tmp2m-34w__ccsm3`, its missing values can be computed through first multiplying the average forecast `nmme-tmp2m-34w__nmmemean` by 9 to get the sum of **nmme-tmp2m-34w** forecasts and then subtract the other 8 forecasts (of 8 remaining weather forecast models).

We will use this methodology to fill missing values for all columns containing missing values. 


In [ ]:
forecast_na_cols = ['nmme0-tmp2m-34w__ccsm30',
 'nmme-tmp2m-56w__ccsm3',
 'nmme-prate-34w__ccsm3',
 'nmme0-prate-56w__ccsm30',
 'nmme0-prate-34w__ccsm30',
 'nmme-prate-56w__ccsm3',
 'nmme-tmp2m-34w__ccsm3']


forecast_mean_cols = ['nmme0-tmp2m-34w__nmme0mean', 
 'nmme-tmp2m-56w__nmmemean', 
 'nmme-prate-34w__nmmemean', 
 'nmme0-prate-56w__nmme0mean', 
 'nmme0-prate-34w__nmme0mean', 
 'nmme-prate-56w__nmmemean', 
 'nmme-tmp2m-34w__nmmemean']


nmme0_tmp2m_34w = ['nmme0-tmp2m-34w__cancm30',
'nmme0-tmp2m-34w__cancm40',
'nmme0-tmp2m-34w__ccsm40',
'nmme0-tmp2m-34w__cfsv20',
'nmme0-tmp2m-34w__gfdlflora0',
'nmme0-tmp2m-34w__gfdlflorb0',
'nmme0-tmp2m-34w__gfdl0',
'nmme0-tmp2m-34w__nasa0']

nmme_tmp2m_56w = ['nmme-tmp2m-56w__cancm3',
'nmme-tmp2m-56w__cancm4',
'nmme-tmp2m-56w__ccsm4',
'nmme-tmp2m-56w__cfsv2',
'nmme-tmp2m-56w__gfdl',
'nmme-tmp2m-56w__gfdlflora',
'nmme-tmp2m-56w__gfdlflorb',
'nmme-tmp2m-56w__nasa']

nmme_prate_34w = ['nmme-prate-34w__cancm3',
'nmme-prate-34w__cancm4',
'nmme-prate-34w__ccsm4',
'nmme-prate-34w__cfsv2',
'nmme-prate-34w__gfdl',
'nmme-prate-34w__gfdlflora',
'nmme-prate-34w__gfdlflorb',
'nmme-prate-34w__nasa']

nmme0_prate_56w = [ 'nmme0-prate-56w__cancm30',
'nmme0-prate-56w__cancm40',
'nmme0-prate-56w__ccsm40',
'nmme0-prate-56w__cfsv20',
'nmme0-prate-56w__gfdlflora0',
'nmme0-prate-56w__gfdlflorb0',
'nmme0-prate-56w__gfdl0',
'nmme0-prate-56w__nasa0']

nmme0_prate_34w = ['nmme0-prate-34w__cancm30',
'nmme0-prate-34w__cancm40',
'nmme0-prate-34w__ccsm40',
'nmme0-prate-34w__cfsv20',
'nmme0-prate-34w__gfdlflora0',
'nmme0-prate-34w__gfdlflorb0',
'nmme0-prate-34w__gfdl0',
'nmme0-prate-34w__nasa0']

nmme_prate_56w = ['nmme-prate-56w__cancm3',
'nmme-prate-56w__cancm4',
'nmme-prate-56w__ccsm4',
'nmme-prate-56w__cfsv2',
'nmme-prate-56w__gfdl',
'nmme-prate-56w__gfdlflora',
'nmme-prate-56w__gfdlflorb',
'nmme-prate-56w__nasa']

nmme_tmp2m_34w = ['nmme-tmp2m-34w__cancm3',
'nmme-tmp2m-34w__cancm4',
'nmme-tmp2m-34w__ccsm4',
'nmme-tmp2m-34w__cfsv2',
'nmme-tmp2m-34w__gfdl',
'nmme-tmp2m-34w__gfdlflora',
'nmme-tmp2m-34w__gfdlflorb',
'nmme-tmp2m-34w__nasa']

forecast_cols = [nmme0_tmp2m_34w, nmme_tmp2m_56w, nmme_prate_34w, nmme0_prate_56w, nmme0_prate_34w, nmme_prate_56w, nmme_tmp2m_34w]

In [ ]:
def impute_missing_values(df):

    for i, na_col in enumerate(forecast_na_cols):
        total = df[forecast_mean_cols[i]] * 9
        sum_of_8_forecast = df[forecast_cols[i]].sum(axis=1)
        df[na_col] = total - sum_of_8_forecast
        
    return df

In [ ]:
df = impute_missing_values(df)

We will just drop the `ccsm30` column as it's redundant.


In [ ]:
def drop_redundant_col(df):
    
    df.drop(['ccsm30'], axis=1, inplace=True)
    
    return df

In [ ]:
df = drop_redundant_col(df)

Now we no longer have missing values.


In [ ]:
df.isnull().sum().sum()

## Feature Selection and Lasso Regressions (to Obtain Residuals)


Boosting algorithms are normally used to predict the residuals of the base models' predictions. To obtain these residuals, we can first fit linear regressions to the data with some regularization. To account for the fact that different locations might have different relationships between its predictors and its response variable, we fit 514 different linear regressions and collect their residuals.


### Train, Validation, Test split


We sort the data used for fitting linear regressions by location group and start date. We split the data so that for each location group, data between 2014-09-1 and 2015-11-15 are used for training, data between 2015-11-15 and 2016-04-08 are used for validation, and data between 2016-04-08 and 2016-08-31 are used for testing. By doing so, we maintain a **6:2:2** ratio of train, validation, and test split.


In [ ]:
df.sort_values(by=['loc_group', 'startdate'], inplace=True, ignore_index=True)

In [ ]:
df.head()

In [ ]:
train = df[df["startdate"] <= pd.datetime(2015, 11, 15)]
val = df[(df["startdate"] > pd.datetime(2015, 11, 15)) & (df["startdate"] <= pd.datetime(2016, 4, 8))]
test = df[df["startdate"] > pd.datetime(2016, 4, 8)]

print(f'Train_shape: {train.shape}  |  Val_shape: {val.shape}   |  Test_shape: {test.shape}')

### Feature Selection For Linear Regression


Since we are fitting linear regression models, predictors that have high linear correlation with the response variable will give us better model fitting results. Thus, we select some features based on this criterion, hoping that the selected features could explain any underlying linear relationships.

The first set of potential features we will look at are the ones whose names contain "14d", which indicates that they are measured over the same time frame as the target variable and thus they have higher chance of being linearly correlated with the target.


In [ ]:
LR_cols = []

for col in df.columns:
    if '14d' in col and col != target:
        corr = df[[col, target]].corr().iloc[0,1]
        if np.abs(corr) > 0.8: 
            print(col, corr)
            LR_cols.append(col)

By setting 0.8 as the threshold, we identified three features with high linear correlations with the target.


Another potential set of features are the NMME model forecasts as they are purely physics-based models which forecast weather-related information such as precipitation. Hence, they have the potential of providing us with some meaningful insights. We have the following prefixes, each of them is associated with the NMME model forecasts over a specific time frame:


In [ ]:
prefixs = []

for col in df.columns:
    if '34w' in col or '56w' in col:
        prefix = col.split("__")[0]
        prefixs.append(prefix)

prefixs = list(set(prefixs))
prefixs

Since there are many columns sharing the same prefix, let's see if information contained in columns sharing a prefix can be "summarized" by linear PCA.


In [ ]:
pca = PCA(n_components=1)
y = np.array(df[target]).reshape(1, -1)

for prefix in prefixs:
    
    cols = [col for col in df.columns if prefix in col and 'mean' not in col]
    pca1 = pca.fit_transform(df[cols]).reshape(1, -1)
    print(prefix, f"| pca1 explained variance: {pca.explained_variance_ratio_[0]}")
    print(f"pca1 vs target corr: {np.corrcoef(pca1, y)[1,0]}")
    print()

From the above result, we can conclude that for "nmme0-tmp2m-34", "nmme-tmp2m-56w", and "nmme-tmp2m-34w" forecasts, the contained information can be summarized by the first principal component and the first principal component is highly correlated with the target (|correlation| > 0.8). 


In [ ]:
corr_prefix = ['nmme0-tmp2m-34w', 'nmme-tmp2m-34w', 'nmme-tmp2m-56w']

corr_prefix_cols = [nmme0_tmp2m_34w, nmme_tmp2m_34w, nmme_tmp2m_56w]
    
for prefix, cols in zip(corr_prefix, corr_prefix_cols):
    pc_name = prefix + '-pc'
    pca = PCA(n_components=1)
    pca.fit(df[cols])
    for ds in [train, val, test]:
        ds[pc_name] = pca.fit_transform(ds[cols])

In [ ]:
LR_cols = LR_cols + [prefix+'-pc' for prefix in corr_prefix]
LR_cols

Now we have 6 continuous variables for fitting linear regressions, we can examine the remaining variables and select them as features if they have high linear correlations with our target.


In [ ]:
for col in df.columns:
    
    if df[col].dtype == 'float64' and col not in LR_cols and 'nmme0-tmp2m-34w' not in col and 'nmme-tmp2m-34w' not in col and 'nmme-tmp2m-56w' not in col:      
        corr = df[[col, target]].corr().iloc[0,1]
        if np.abs(corr) > 0.8:
            LR_cols.append(col)
            print(col, corr)

We now have 10 continuous predictors for fitting linear regressions. They are:


In [ ]:
LR_cols = [col for col in LR_cols if col != target]
LR_cols

### Selecting Alpha and Fitting Linear Regressions (to Obtain Residuals)


We use the following defined function `LR_models` to help us fit 514 linear regressions to the weather data of 514 different locations. Notice that the function takes three parameters:

- `alpha`: float, regularization parameter shared by the 514 linear regressions

- `selected_feats`: list of strings, selected features for fitting linear regressions to the data

- `shift_feats`: boolean, True if we want to shift the `..-34w-..` and `..-56w-..` features backward by 14 and 28 days respectively, by location group.

The function will return the training rmse and validation rmse.


In [ ]:
def LR_models(alpha, selected_feats, shift_feats):
    
    # new train and vaidation dataset containing only the selected features
    X_train, y_train = train[selected_feats], train[target]
    X_val, y_val = val[selected_feats], val[target]

    train_se = 0
    val_se = 0

    # loop through 514 different locations, 
    # for each location we create a machine learning pipeline and fit a linear regression model
    for i in range(df.loc_group.nunique()):

        # train and validation dataset for the ith location
        X1_train = X_train[train["loc_group"]==i]
        y1_train = y_train[train["loc_group"]==i]
        X1_val = X_val[val["loc_group"]==i]
        y1_val = y_val[val["loc_group"]==i]

        # collect the indices of the observations, which we will use to recover the train and val split of the data
        t_index = X_train[train["loc_group"]==i].index
        v_index = X_val[val["loc_group"]==i].index

        # if shift_feats, we combine the train and val set to shift the "34w" and "56w" features 
        if shift_feats == True and np.any(['34w' in col or '56w' in col for col in selected_feats]):
            X1_conc = pd.concat([X1_train, X1_val])           
            for col in selected_feats:
                if '34w' in col:
                    X1_conc[col] = X1_conc[col].shift(-14).ffill()
                if '56w' in col:
                    X1_conc[col] = X1_conc[col].shift(-28).ffill()

            # we recover the train and val set by their respective list of indices
            X1_train = X1_conc.loc[t_index, :]
            X1_val = X1_conc.loc[v_index, :]
        
        # scale the data
        ss = StandardScaler()
        X1_train_ss = ss.fit_transform(X1_train)
        X1_val_ss = ss.transform(X1_val)

        # fit the linear regression model with alpha as the regularization parameter
        LR = Lasso(alpha=alpha, max_iter=10000)
        LR.fit(X1_train_ss, y1_train)
        y1_train_pred = LR.predict(X1_train_ss)
        y1_val_pred = LR.predict(X1_val_ss)
        
        # accumulate the total squared errors by the sum of squared errors from the ith location
        train_se += mean_squared_error(y1_train, y1_train_pred) * len(t_index)
        val_se += mean_squared_error(y1_val, y1_val_pred) * len(v_index)
        
    # return the root mean squared error of train and validation set
    train_rmse = np.round(np.sqrt(train_se / X_train.shape[0]), 3)
    val_rmse = np.round(np.sqrt(val_se / X_val.shape[0]), 3)
    
    return train_rmse, val_rmse

One important parameter in fitting Linear Regressions is the `alpha` parameter. Here we are going to try a list of different alphas and then look at the root mean squared error of the training and validation dataset. 


In [ ]:
train_rmses = []
val_rmses = []
alphas = [0.001, 0.01, 0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.5, 0.8, 1, 1.5, 2, 2.5, 3, 3.5, 4]

for alpha in alphas:
    
    print(f"alpha = {alpha}...")
    train_rmse, val_rmse = LR_models(alpha, LR_cols, True)
    train_rmses.append(train_rmse)
    val_rmses.append(val_rmse)

In [ ]:
plt.plot(alphas, train_rmses, 'o-', label="train rmse")
plt.plot(alphas, val_rmses,'o-', label="val rmse")
plt.plot(alphas, np.array(val_rmses)-np.array(train_rmses), 'o-', label="val-train rmse")

for i in np.arange(8, len(alphas)):
    plt.annotate(f"{np.round(val_rmses[i]-train_rmses[i], 2)}", (alphas[i], val_rmses[i]-train_rmses[i]))

plt.xlabel("lasso regression alpha")
plt.ylabel("rmse")
plt.legend(loc='best')
plt.show()

We want the error value to be as small as possible but we also don't want the models to overfit to the training data. Thus, to take both the error metric and the overfitting problem into account, we decide to use **alpha = 2** to create the linear regression residuals.


To obtain the residuals from the linear regressions, we are going to modify the `LR_models` function. Note that we don't need the `alpha` argument in the function anymore as it's now a fixed value.

To have more training data to learn the 514 linear regressions, we are going to combine `train` and `val` into one new training dataset and we call it `train2`.


In [ ]:
train2 = pd.concat([train, val], axis=0)
print(f'Train2_shape: {train2.shape}  |  Test_shape: {test.shape}')

Here is the modified function for obtaining linear regression residuals. It will return the fitted residuals as well as the predictions for the entire dataset.


In [ ]:
def LR_models_residuals(selected_feats, shift_feats):
    
    # new train and test dataset containing only the selected features
    X_train, y_train = train2[selected_feats], train2[target]
    X_test, y_test = test[selected_feats], test[target]

    pred_residuals = np.zeros(df.shape[0])
    predictions = np.zeros(df.shape[0])

    # loop through 514 different locations, 
    # for each location we create a machine learning pipeline and fit a linear regression model
    for i in range(df.loc_group.nunique()):

        # train and test dataset for the ith location
        X1_train = X_train[train2["loc_group"]==i]
        y1_train = y_train[train2["loc_group"]==i]
        X1_test = X_test[test["loc_group"]==i]
        y1_test = y_test[test["loc_group"]==i]

        # collect the indices of the observations, which we will use to recover the train and test split of the data
        t_index = X_train[train2["loc_group"]==i].index
        te_index = X_test[test["loc_group"]==i].index

        # if shift_feats, we combine the train and test set to shift the "34w" and "56w" features 
        if shift_feats == True and np.any(['34w' in col or '56w' in col for col in selected_feats]):
            X1_conc = pd.concat([X1_train, X1_test])           
            for col in selected_feats:
                if '34w' in col:
                    X1_conc[col] = X1_conc[col].shift(-14).ffill()
                if '56w' in col:
                    X1_conc[col] = X1_conc[col].shift(-28).ffill()

            # we recover the train and test set by their respective list of indices
            X1_train = X1_conc.loc[t_index, :]
            X1_test = X1_conc.loc[te_index, :]
        
        # scale the data
        ss = StandardScaler()
        X1_train_ss = ss.fit_transform(X1_train)
        X1_test_ss = ss.transform(X1_test)

        # fit the linear regression model with alpha = 2.0 as the regularization parameter
        LR = Lasso(alpha=2.0, max_iter=10000)
        LR.fit(X1_train_ss, y1_train)
        y1_train_pred = LR.predict(X1_train_ss)
        y1_test_pred = LR.predict(X1_test_ss)
        predictions[t_index] = y1_train_pred
        predictions[te_index] = y1_test_pred
        
        # return predicted residuals
        pred_residuals[t_index] = y1_train - y1_train_pred
        pred_residuals[te_index] = y1_test - y1_test_pred
    
    return pred_residuals, predictions

In [ ]:
pred_residuals, predictions = LR_models_residuals(LR_cols, True)

We confirm that we have obtained linear regression fitted residuals for all observations in the original dataset.


In [ ]:
print(len(pred_residuals) == df.shape[0])

We can plot a histogram of the fitted residuals:


In [ ]:
plt.hist(pred_residuals, density=True)
plt.xlabel("Fitted residuals")
plt.ylabel("Normalized count")
plt.title("Lassion regression fitted residuals")
plt.show()

## Using Boosting Algorithms to Predict Residuals

CatBoost, LightGBM, and XGBoost are three popular boosting algorithms in machine learning. Before we dive into the these three algorithms, let's start with the term ensemble learning and how boosting as one type of ensemble learning helps us reduce model errors.

**Ensemble Learning** is a Machine Learning technique that combines predictions from multiple base models to produce more accurate and robust predictions. Ensemble learning can improve the generalization of the model. It can help to create a model that can generalize well to new and unseen data by combining the different perspectives of individual models. There are several ways to create an ensemble model, including **bagging, boosting, and stacking**. 

**Bagging** is primarily used to reduce the variance of a model by combining predictions of multiple models that are trained on different subsets of the training data through a voting mechanism. **Boosting**, on the other hand, is used to reduce the bias of a model by sequentially training models that focus on correcting the errors of the previous models. **Stacking** is used to combine multiple models by training a meta-model on the predictions of the base models to create more accurate final predictions.

In this guided project, **we will focus on the Boosting algorithms**, which are particularly helpful for achieving success in data science competitions.


First, we concatenate the fitted residuals to the dataframe and make sure that there are no null values in the `residuals` column as it will be our target column when fitting the Boosting models.


In [ ]:
df['residuals'] = pred_residuals
df.residuals.isnull().sum()

In [ ]:
del pred_residuals

We add the `residuals` column to the training and test dataset as well.


In [ ]:
train2['residuals'] = df.loc[train2.index, 'residuals']
test['residuals'] = df.loc[test.index, 'residuals']

### Feature Selection for Boosting


Now we need to select features for the Boosting models, which we will call `B_cols`. First, we will use the forecasts from the purely physics-based weather models and we will select the forecast means this time.


In [ ]:
forecast_means = [col for col in df.columns if 'mean' in col]
forecast_means

As we did before with linear regressions, we are going to shift the `..-34w-..` and `..-56w-..` features backward by 14 and 28 days respectively, by location group, and then we add those shifted features to `B_cols`.


In [ ]:
B_cols = []

for col in forecast_means:
    if '34w' in col:
        df[col+'_shifted'] = df.groupby(['loc_group'])[col].shift(-14).ffill()
        B_cols.append(col+'_shifted')
    elif '56w' in col:
        df[col+'_shifted'] = df.groupby(['loc_group'])[col].shift(-28).ffill()
        B_cols.append(col+'_shifted')   
    else:   
        B_cols.append(col)

Now we append these shifted features to our training and test dataset.


In [ ]:
for ds in [train2, test]:
    for col in B_cols:
        ds[col] = df.loc[ds.index, col]

print(train2.shape, test.shape)

Another set of continuous variables we could add to `B_cols` are the ones whose names contain "14d", because they are measured over the same time interval as our original target variable.


In [ ]:
for col in df.columns:
    if '14d' in col:
        B_cols.append(col)

Apart from the continuous variables, don't forget about the categorical variables that we haven't touched on so far. The original dataset comes with one categorical variable called `climateregions__climateregion`. We can encode this variable using Label Encoder from scikit-learn. **Note that we can analyze the feature importance later after we fit the models**.


In [ ]:
LE = LabelEncoder()
LE.fit(train2['climateregions__climateregion'])
train2['climateregions__climateregion'] = LE.transform(train2['climateregions__climateregion'])
test['climateregions__climateregion'] = LE.transform(test['climateregions__climateregion'])

We can create a `month` variable to see if some months have larger fitted residuals than the others.


In [ ]:
df['month'] = df['startdate'].dt.month
df.groupby(['month'])['residuals'].median()

It appears that our linear regression fitted residuals differ quite perceivably for different months, so we could consider `month` as a feature for the Boosting models. Let's create `month` variable for the training and test set.


In [ ]:
for ds in [train2, test]:
    ds['month'] = ds['startdate'].dt.month

`loc_group` is also an important categorical variable as it specifies the location of the weather data. According to the competition, `elevation__elevation` is also an important feature, so we will add them to `B_cols`.


In [ ]:
B_cols += ['climateregions__climateregion', 'month', 'loc_group', 'elevation__elevation']

print(f"We have {len(B_cols)} features for fitting the Boosting models.")
B_cols

In [ ]:
del ds

### Boosting Models for Predicting Fitted Residuals


As we are no longer looking to explain any linear relationships between the predictors and the target variable (as they have been handled by the linear regressions we fitted), we will move on to using some Boosting models includeing CatBoost, LightGBM, and XGBoost to capture complex non-linear relationships between the input features and the fitted residuals. 

We will use a combination of three Boosting models to predict our fitted residuals, so basically we are trying to see if there is any non-linear pattern in the fitted residuals that we could explain using Boosting models. If the patterm we found is generalizable to unseen data, we can further reduce the gap between what Machine Learning models can predict and the ground truths. 

- **CatBoost**: CatBoost works by iteratively building a set of decision trees. At each iteration, it calculates the gradient of the loss function with respect to the current set of predictions, and fits a decision tree to the negative gradient. The predictions from the new decision tree are then added to the current predictions, and the process is repeated for a specified number of iterations. CatBoost also incorporates a unique mechanism for handling categorical variables, which it achieves by using an ordered boosting technique that sorts the categories in a way that optimizes the loss function.

- **LightGBM**: LightGBM, on the other hand, uses a different approach for building decision trees. Instead of using a depth-first approach like many other Boosting models, LightGBM uses a leaf-wise approach that grows the tree by expanding the leaf with the largest loss reduction. This approach can be more efficient than depth-first approaches because it reduces the number of data points that need to be processed for each split. LightGBM also uses the Gradient-based One-Side Sampling (GOSS) technique to sample the most important instances for each iteration, which further improves efficiency.

- **XGBoost**: XGBoost also builds decision trees, but it uses a slightly different algorithm. Like CatBoost and LightGBM, XGBoost uses a gradient-based approach to build decision trees. However, it also includes several advanced features such as regularization techniques, early stopping, and a second-order approximation of the gradient to improve prediction accuracy. XGBoost also allows for parallel processing, which can be useful for large-scale datasets.


<center><img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-GPXX0CRSEN/images/1_V-vikG-ye2Y03R943d0CPw.jpeg" width="50%"></center>

<p style="color:gray; text-align:center;">Credit: enjoyalgorithms.com</p>


We need to specify catrgorical columns for the three Boosting models.


In [ ]:
cat_cols =  ['climateregions__climateregion', 'month', 'loc_group']

We use the `boosting` function to help us fit three Boosting models at one. The function takes in three parameter:

- `ratios`: list of floats, specifying the weighting of each model in the final predictions.

- `selected_cols`: list of strings, selected features for fitting the three Boosting models.

- `cat_cols`: list of strings, names of categorical columns in `selected_cols`.

The function will return the predicted residuals for the training and test set.


In [ ]:
def boosting(ratios=[0.6, 0.3, 0.1], selected_cols=B_cols, cat_cols=cat_cols):
    
    
    X_train, y_train = train2[selected_cols], train2['residuals']
    X_test, y_test = test[selected_cols], test['residuals']

    for ds in [X_train, X_test]:
        ds['loc_group'] = ds['loc_group'].astype("category")
        ds['climateregions__climateregion'] = ds['climateregions__climateregion'].astype("category")
        ds['month'] = ds['month'].astype("category")


    cat_inds = [selected_cols.index(col) for col in cat_cols]

    train_pred_cb, train_pred_lgb, train_pred_xgb = [np.zeros(len(y_train))] * 3
    test_pred_cb, test_pred_lgb, test_pred_xgb = [np.zeros(len(y_test))] * 3

    if ratios[0] > 0:
        cbr = cb.CatBoostRegressor(loss_function="RMSE", random_state=42, verbose=1)  
        cbr.fit(X_train, y_train, cat_features=cat_inds)
        train_pred_cb = cbr.predict(X_train)
        test_pred_cb = cbr.predict(X_test)
    if ratios[1] > 0:
        lgbr = lgb.LGBMRegressor(objective='regression', metric='rmse', random_state=42, verbose=1)       
        lgbr.fit(X_train, y_train, categorical_feature=cat_inds)
        train_pred_lgb = lgbr.predict(X_train)
        test_pred_lgb = lgbr.predict(X_test)
    if ratios[2] > 0:
        xgbr = xgb.XGBRegressor(objective="reg:squarederror", eval_metric='rmse', enable_categorical=True, random_state=42, verbose=0)
        xgbr.fit(X_train, y_train)
        train_pred_xgb = xgbr.predict(X_train)
        test_pred_xgb = xgbr.predict(X_test)

    train_pred = ratios[0] * train_pred_cb + ratios[1] * train_pred_lgb + ratios[2] * train_pred_xgb
    test_pred = ratios[0] * test_pred_cb + ratios[1] * test_pred_lgb + ratios[2] * test_pred_xgb

    print(f"train rmse = {rmse(y_train, train_pred)}, test rmse = {rmse(y_test, test_pred)}")
    
    return train_pred, test_pred

The predicted residuals are the part of the fitted residuals that could be captured by the complex non-linear Boosting models. The difference between the predicted residuals and our linear regression fitted residuals will be the final error of our predictions.

In the next cell, you can alter the `ratios` argument to increase/decrease each model's contribution to the final prediction and check the resulting training and test root mean squared error. 

As there is a memory limit with this lab and our dataset is quite large, LightGBM and XGBoost training are commented out. According to experience, CatBoost tends to give the best performance. If you have local or cloud training resources, feel free to download this notebook and play with the `ratios` argument.


Note: The following CatBoost training is expected to take approximately 5 minutes.


In [ ]:
print("Pure CatBoost:")
train_pred1, test_pred1 = boosting(ratios=[1, 0, 0])
# print("Pure LightGBM:")
# train_pred2, test_pred2 = boosting(ratios=[0, 1, 0])
# print("Pure XGBoost:")
# train_pred3, test_pred3 = boosting(ratios=[0, 0, 1])

### Plotting Predictions


As we have now obtained the residuals predicted by the Boosting models(s). We can add them back to the predictions by the Lasso Regressions to obtain the final target predictions.


We append linear predictions, which were produced by the 514 Lasso Regressions, to our original data frame:


In [ ]:
df['linearPrediction'] = predictions 

In [ ]:
del predictions

We append the residuals predicted by the Boosting model(s) to the original data frame as well.


In [ ]:
train2_inds = train2.index
test_inds = test.index

In [ ]:
df['boostedResiduals'] = 0
df.loc[train2_inds, 'boostedResiduals'] = train_pred1
df.loc[test_inds, 'boostedResiduals'] = test_pred1

In [ ]:
del train2_inds
del test_inds

In [ ]:
plt.hist(df['boostedResiduals'], density=True)
plt.xlabel("Boosted residuals")
plt.ylabel("Normalized count")
plt.title("Boosting fitted residuals")
plt.show()

We confirm that there are null values in the boosted residuals and the linear predictions.


In [ ]:
df.boostedResiduals.isnull().sum(), df.linearPrediction.isnull().sum()

Final predictions equal linear predictions plus boosted residuals.


In [ ]:
df['finalPrediction'] = df['linearPrediction'] + df['boostedResiduals']

Since it would be hard to visualize predictions for all 514 locations, we will pick the locations whose forecasts have the smallest root mean squared error (rmse) and we visualize those locations.


We calculate the rmse of all the groups and store the calculations in `boostedresiduals_dict`:


In [ ]:
boostedresiduals_grouped = df.groupby(['loc_group'])
boostedresiduals_dict = {}

for name, group in boostedresiduals_grouped:
    
    rmse = np.sqrt(np.sum(group['boostedResiduals']**2) / group.shape[0])
    boostedresiduals_dict[name] = rmse

In [ ]:
del group

We sort the location groups by the corresponding rmses in ascending order and we pick four best predicted groups:


In [ ]:
sorted_boostedresiduals_dict = dict(sorted(boostedresiduals_dict.items(), key=lambda item: item[1]))
best_predicted_groups = list(sorted_boostedresiduals_dict.keys())[:4]
best_predicted_groups

We plot the final predictions:


In [ ]:
plt.figure(figsize=(12, 6))

for i, group in enumerate(best_predicted_groups):
    
    plt.subplot(2,2,i+1)
    train2_inds = train2.loc[train2.loc_group == group].index
    test_inds = test.loc[test.loc_group == group].index
    
    plt.plot(df.loc[df.loc_group == group, 'startdate'], df.loc[df.loc_group == group, target], label='target')
    plt.plot(df.loc[test_inds, 'startdate'], df.loc[test_inds, 'finalPrediction'], label='predictions')
    
    plt.title(f"Group{group} predictions")
    plt.xticks(rotation=45)
    plt.legend(loc='lower right')
    
plt.suptitle("Groups that have the smallest average rmse")
plt.tight_layout()

## Congratulations! You have completed this guided project.


This guided project demonstrates one way of jump-starting your Data Science competition. Data Science or Machine Learning projects should always follow the following cycle:

- **Problem formulation**: defining the problem and understanding the business context.

- **Data collection**: gathering and acquiring data from various sources.

- **Data preprocessing**: cleaning, transforming, and preparing data for analysis.

- **Exploratory data analysis**: understanding the structure, contents, and patterns in the data through visualizations and summary statistics.

- **Feature engineering**: creating new features from the existing data.

- **Model selection**: choosing an appropriate model or algorithm for the problem.

- **Model training**: fitting the model to the data and tuning its hyperparameters.

- **Model evaluation**: assessing the performance of the model using appropriate metrics and techniques.


However, please keep in mind that at each step of the cycle, there exists a million of different solutions and tools for solving the problem. It's a rapidly evolving field, so there is on one fixed set of winning techniques, but as long as you keep learning and practicing you will eventually develop your own set of rules for your Data Science game! Happy Learning!


## Author(s)


[Roxanne Li](https://www.linkedin.com/in/roxanne-li/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkGuidedProjectsIBMGPXX0CRSEN2517-2023-01-01)


### Other Contributor(s)


[Sina Nazeri](https://www.linkedin.com/in/sina-nazeri?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkGuidedProjectsIBMGPXX0CRSEN2517-2023-01-01)  
[JC Chen](contributor_link)


## Change Log


|Date (YYYY-MM-DD)|Version|Changed By|Change Description|
|-|-|-|-|
|2020-07-17|0.1|Sam|Create Lab Template|
|2023-04-03|0.1|Roxanne Li|Create Project First Draft|
|2023-04-06|0.1|JC Chen|Reviewed Project First Draft|
|2023-04-15|0.2|Roxanne Li|Updated Project|


Copyright © 2022 IBM Corporation. All rights reserved.
